src/mame/sony_news/cxd8442q.cpp

```diff
@@ -0,0 +1,364 @@
+	// It isn't 100% clear what the trigger for clearing intstat is
+	// but this seems like a reasonable place for it to go, and it works.
+	fifo_channels[channel].intstat = 0x0;
+	fifo_channels[channel].intctrl = data;
+	irq_check();
+}
+
+uint32_t cxd8442q_device::read_intstat(offs_t offset)
+{
+	uint32_t channel = dw_offset_to_channel(offset);
+	// There seems to be more in this register, but this is the minimum to get the ESCCF working.
+	auto intstat = fifo_channels[channel].intstat;
+	auto mask = fifo_channels[channel].intctrl & 0x1;
+	return intstat & mask;
+}
+
+uint32_t cxd8442q_device::read_count(offs_t offset)
+{
+	return fifo_channels[dw_offset_to_channel(offset)].count;
+}
+
+uint8_t cxd8442q_device::read_fifo_data(offs_t offset)
+{
+	return fifo_channels[byte_offset_to_channel(offset)].read_data_from_fifo();
+}
+
+void cxd8442q_device::write_fifo_data(offs_t offset, uint8_t data)
+{
+	uint32_t channel = byte_offset_to_channel(offset);
+	LOG("FIFO CH%d: Pushing 0x%x\n", channel, data, machine().describe_context());
+	fifo_channels[channel].write_data_to_fifo(data);
+}
+
+uint32_t cxd8442q_device::read_fifo_ram(offs_t offset)
+{
+	return fifo_ram[offset];
+}
+
+void cxd8442q_device::write_fifo_ram(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	COMBINE_DATA(&fifo_ram[offset]);
+}
+
+void cxd8442q_device::device_start()
+{
+	fifo_ram = std::make_unique<uint32_t[]>(FIFO_MAX_RAM_SIZE);
+	fifo_timer = machine().scheduler().timer_alloc(timer_expired_delegate(FUNC(cxd8442q_device::fifo_dma_execute), this));
+
+	for (int channel = 0; channel < FIFO_CH_TOTAL; ++channel)
+	{
+		fifo_channels[channel].resolve_callbacks();
+	}
+
+	save_pointer(NAME(fifo_ram), FIFO_MAX_RAM_SIZE);
+	save_item(STRUCT_MEMBER(fifo_channels, fifo_size));
+	save_item(STRUCT_MEMBER(fifo_channels, address));
+	save_item(STRUCT_MEMBER(fifo_channels, dma_mode));
+	save_item(STRUCT_MEMBER(fifo_channels, intctrl));
+	save_item(STRUCT_MEMBER(fifo_channels, intstat));
+	save_item(STRUCT_MEMBER(fifo_channels, count));
+	save_item(STRUCT_MEMBER(fifo_channels, drq));
+	save_item(STRUCT_MEMBER(fifo_channels, fifo_w_position));
+	save_item(STRUCT_MEMBER(fifo_channels, fifo_r_position));
+}
+
+void cxd8442q_device::device_reset()
+{
+	for (int channel = 0; channel < FIFO_CH_TOTAL; ++channel)
+	{
+		fifo_channels[channel].reset();
+	}
+}
+
+TIMER_CALLBACK_MEMBER(cxd8442q_device::fifo_dma_execute)
+{
+	bool dma_active = false;
+	for (int channel = 0; channel < FIFO_CH_TOTAL; ++channel)
+	{
+		apfifo_channel &this_channel = fifo_channels[channel];
+
+		// Skip this channel if we don't need to do anything
+		if (!(this_channel.dma_mode & apfifo_channel::DMA_EN))
+		{
+			continue;
+		}
+
+		// Check DRQ to see if the device is ready to give or receive data
+		if (this_channel.drq)
+		{
+			if (this_channel.dma_cycle())
+			{
+				dma_active = true;
+			}
+		}
+		else
+		{
+			dma_active = true;
+		}
+	}
```

Comment:
This could probably use a range-based `for` loop.  Also, would the body be better off being encapsulated as a member function of the channel structure?  It could return a `bool` indicating whether its DMA is active.

Reply:
Yeah, that works better. Updated.